### PR TITLE
Add optional caliban endpoint

### LINKED ISSUE #5: Add optional GraphQL endpoint
Use Caliban

### PR DESCRIPTION
**Pull request by vpavkin**
Fix #5 

For some reason I bumped into https://github.com/foundweekends/giter8/pull/432, although giter plugin version is up to date.

The only consequence so far is that there's an unneeded html file in resources folder if user opts out of caliban endpoint

### CODE REVIEW

File: src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$ [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      maxDepth(30) @@                 // query analyzer that limit query depth
+      timeout(3 seconds) @@           // wrapper that fails slow queries
+      printSlowQueries(500 millis) @@ // wrapper that logs slow queries
+      apolloTracing                   // wrapper for https://github.com/apollographql/apollo-tracing
+
+  val live: ZLayer[ItemRepository with Has[ActorSystem], CalibanError.ValidationError, GraphQLApi] =
+    ZLayer.fromFunctionM { env =>
+      implicit val system = env.get[ActorSystem]
+      implicit val ec     = system.dispatcher
+      val adapter = AkkaHttpAdapter(new PlayJsonBackend)
```
Comment (marioosh) [inline src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$:71]: why PlayJsonBackend is in Akka-Http ? Do we need `Play` here? Or is this just JSON parser?

**Comment by marioosh** [comment]
Does it work?? Actually I'm missing one part (maybe it's automagically covered by caliban somehow - I don't know, don't know the library so far.... :D
JSON sent to the endpoint is like this:
```
 { operationName: "", query: {}, variables: {} } 
```
Query is passed to the executor. I don't where it's parsed, so the reason of my question. Are you able to run server with console and make a query?

File: src/main/g8/src/main/scala/$package$/Boot.scala [hunk 2]
```diff
@@ -2,36 +2,47 @@ package $package$
...
-      api                            <- ZIO.access[Api](_.get)
-      binding                        <- ZIO.fromFuture(_ => Http().bindAndHandle(api.routes, cfg.host, cfg.port))
-      _                              <- putStrLn(s"Server online at http://\${cfg.host}:\${cfg.port}/\nPress RETURN to stop...")
-      _                              <- getStrLn
-      _                              <- ZIO.fromFuture(_ => binding.unbind())
-      _                              <- ZIO.fromFuture(_ => system.terminate())
-    } yield ()
-  }.flatten
+  val program
+    : ZIO[Console with Api $if(add_caliban_endpoint.truthy)$with GraphQLApi $endif$with Has[ActorSystem] with Has[ActorSystem] with Config[ApiConfig], Throwable, Unit] =
```
[the remark this answers is not in the document]
Comment (vpavkin) [reply src/main/g8/src/main/scala/$package$/Boot.scala:23]: Ooops :) Will fix, thanks!

File: src/main/g8/src/main/scala/$package$/Boot.scala [hunk 3]
```diff
@@ -2,36 +2,47 @@ package $package$
...
+        _                              <- putStrLn(s"Server online at http://\${cfg.host}:\${cfg.port}/\nPress RETURN to stop...")
+        _                              <- getStrLn
+        _                              <- ZIO.fromFuture(_ => binding.unbind())
+      } yield ()
+    }.flatten
 
   val loadConfig = ZIO.effect(ConfigFactory.load.resolve)
 
+  val actorSystem = ZLayer.fromManaged(
+    ZManaged.make(ZIO.effect(ActorSystem("zio-example-system")))(s => ZIO.fromFuture(_ => s.terminate()).either)
```
[the remark this answers is not in the document]
Comment (vpavkin) [reply src/main/g8/src/main/scala/$package$/Boot.scala:43]: good suggestion, I'll try to use template variable here 👍

File: src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$ [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      api(env).interpreter.map(interpreter =>
+        new Service {
+          def routes: Route =
+            path("api" / "graphql") {
+              adapter.makeHttpService(interpreter)
+            } ~
+              path("graphiql") {
```
[the remark this answers is not in the document]
Comment (vpavkin) [reply src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$:78]: I remember checking it, but it was long ago, so maybe you're right. will check again.

File: src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$ (hunk 1 above)
Comment (vpavkin) [reply src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$:71]: It's just json parser + encoder. Circe was a default in caliban previously, and it doesn't fit to this stack, so I refactored it a bit. Details here: https://github.com/ghostdogpr/caliban/releases

TLDR - no play backend here, just json :)

**Comment by vpavkin** [comment]
@marioosh WRT how the endpoint works - I'm pretty sure it is, because I tested it with [Altair](https://altair.sirmuel.design/). It even gets the docs properly, so you don't have to type queries by hand.

File: src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$ (hunk 4 above)
[the remark this answers is not in the document]
Comment (vpavkin) [reply src/main/g8/src/main/scala/$package$/api/$if(add_caliban_endpoint.truthy)$GraphQLApi.scala$endif$:78]: just checked, it's `api/graphql` (without `i`), which points exactly to the `path("api" / "graphql")` route defined previously. so should work fine.

**Comment by vpavkin** [comment]
So this is hanging for long time already :) Should I merge it after #33 is merged?

**Comment by jczuchnowski** [comment]
@vpavkin @marioosh sorry I've been delaying it for so long. Looks great. One thing I've been thinking is that in case someone opts out of the GraphQL endpoint, still gets the additional repository and service methods for deletion events, custom search by price etc. Don't you think it would be better if these would also depend on `$if(add_caliban_endpoint.truthy)$` ?

**Comment by vpavkin** [comment]
@jczuchnowski great catch, I missed that entirely. I can clean it up on monday

**Comment by jczuchnowski** [comment]
Don't you have holidays on Monday? :)

**Comment by vpavkin** [comment]
Nope, my holidays start on 1st of May:) Actually finished project a day earlier, so plan to merge this today :)

**Comment by vpavkin** [comment]
@jczuchnowski I added a small cool thing to CI - it now runs tests for both caliban and non-caliban versions of the template. will be handy since we have a lot of conditionals there and it's easy to miss something.

**Comment by vpavkin** [comment]
Cleaned everything, at least it seems so :)
